src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +34,368 @@
+            return InventoryUtils.isPickaxeInTileEntity(worker.getInventoryCitizen(), buildingToDeliver.getNeededPickaxeLevel());
+        }
+        return InventoryUtils.isToolInTileEntity(worker.getInventoryCitizen(), requiredTool);
+    }
+
+    /**
+     * Gather item
+     * @param buildingToDeliver
+     */
+    private void gatherItems(final AbstractBuilding buildingToDeliver)
+    {
+        BlockPos position;
+        if(itemsToDeliver.isEmpty())
+        {
+            String tool = buildingToDeliver.getRequiredTool();
+            position = wareHouse.getTileEntity().getPositionOfChestWithTool(tool, Utils.PICKAXE.equals(tool) ? buildingToDeliver.getNeededPickaxeLevel() : -1);
+        }
+        else
+        {
+            //todo take at least the amount of items he needs.
+            ItemStack stack = itemsToDeliver.get(0);
+            position = wareHouse.getTileEntity().getPositionOfChestWithItemStack(stack);
+        }
+
+        if(position == null)
+        {
+            return;
+        }
+
+        if(!worker.isWorkerAtSiteWithMove(position, MIN_DISTANCE_TO_WAREHOUSE))
+        {
+            return;
+        }
+
+        TileEntity tileEntity = world.getTileEntity(position);
+        if(tileEntity instanceof TileEntityChest)
+        {
+            if (itemsToDeliver.isEmpty())
+            {
+                String tool = buildingToDeliver.getRequiredTool();
+                isToolInTileEntity((TileEntityChest) tileEntity, tool);
+            }
+            else
+            {
+                ItemStack stack = itemsToDeliver.get(0);
+                if(isInTileEntity((TileEntityChest) tileEntity, stack))
+                {
+                    itemsToDeliver.remove(0);
+                }
+            }
+        }
+    }
+
+
+    /**
+     * Check the wareHouse for the next task.
+     * @return the next AiState to go to.
+     */
+    private AIState checkWareHouse()
+    {
+        if(!worker.isWorkerAtSiteWithMove(wareHouse.getLocation(), MIN_DISTANCE_TO_WAREHOUSE))
+        {
+            return START_WORKING;
+        }
+
+        @Nullable final AbstractBuilding buildingToDeliver = wareHouse.getTileEntity().getTask();
+        if(buildingToDeliver == null)
+        {
+            return GATHERING;
+        }
+
+        AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if(ownBuilding instanceof BuildingDeliveryman)
+        {
+            ((BuildingDeliveryman) ownBuilding).setBuildingToDeliver(buildingToDeliver);
+        }
+
+        return PREPARE_DELIVERY;
+    }
+
+    /**
+     * Check if the deliveryman code should be executed.
+     * More concretely if he has a warehouse to work at.
+     * @return false if should continue as planned.
+     */
+    private boolean checkIfExecute()
+    {
+        if(wareHouse != null && wareHouse.getTileEntity() != null)
+        {
+            return false;
+        }
+
+        Map<BlockPos, AbstractBuilding> buildings = job.getColony().getBuildings();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'buildings' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)